### PR TITLE
Testing guide

### LINKED ISSUE #1293: Testing guide
We are looking at providing an overview of what "testing with Prisma" entails:
* Using Prisma migrate to set a testing environment up in a shape for running integration tests (as elaborated on in [this article](https://dev.to/eddeee888/how-to-write-tests-for-prisma-with-docker-and-jest-593i))
* Mocking a client to simplify such a setup when doing unit testing
* Testing an actual migration (e.g. whenever involving custom changes or data transformations)

This article would talk to the various kinds of testing situations where Prisma is involved and offer pointers to existing articles first.
We'd like to write the one about testing an actual migration using the concept of data transformation: custom migration handling combining firstName and lastName into fullName could be something to create a test for, to add confidence in the process of running such a migration on a production environment.

### LINKED ISSUE #5014: Docs: Testing with Jest and Jest.Mock
I am trying to figure out a good pattern for using Jest with Photon. If anyone has suggestions it would be great. And Would probably be good to document this somewhere as well.

I have tried out a couple different things but they don't seem to be working with the `jest.mock` function. Every time I try to run `jest.mock('@prisma/photon')` it doesn't actually mock out the instance that has been created.

### PR DESCRIPTION
**Pull request by molebox**
This PR create sa new `Testing guide` section which has 2 new pages:
- Unit tetsing
- Integration testing

The Unit testing page covers why and when to unit test as well as mocking the client using a singleton and depencency injection patterns, plus a couple of examples of actual unit test.

The Integration testing page covers setting up and using Docker to house a database to test against. It also includes soem examples of integration tests.

Fixes: #1293
Fixes: [5014](https://github.com/prisma/prisma/issues/5014)

### CODE REVIEW

File: content/300-guides/150-testing/100-unit-testing.mdx [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+
+<TopBlock>
+
+Unit testing aims to isolate a small portion (unit) of code and test it for logically predictable behaviors. It generally involves mocking objects or server responses to simulate real world behaviors. Some benefits to unit testing include:
+
+- Quickly find and isolate bugs in code.
+- Provides documentation for each module of code by way of indicating what certain code blocks should be doing.
+- A helpful gauge that a refactor has gone well. The tests should still pass after code has been refactored.
+
+In the context of Prisma, this generally means testing a function which makes database calls using the Prisma client.
```
Comment (mhwelander) [inline content/300-guides/150-testing/100-unit-testing.mdx]:
```suggestion
In the context of Prisma, this generally means testing a function which makes database calls using the Prisma Client.
```

File: content/300-guides/150-testing/100-unit-testing.mdx [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+
+A single test should focus on the data returned by the Prisma client and how the function handles that data, ensuring that it behaves as expected.
+
+This means that you should aim to remove as many dependencies as possible, such as external services and databases, to keep the tests and their environments as lightweight as possible.
+
+</TopBlock>
+
+#### Prerequisites
```
Comment (mhwelander) [inline content/300-guides/150-testing/100-unit-testing.mdx]:
```suggestion
## Prerequisites
```

Creates an h4 - not semantically accurate

File: content/300-guides/150-testing/100-unit-testing.mdx [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+
+Unit testing aims to isolate a small portion (unit) of code and test it for logically predictable behaviors. It generally involves mocking objects or server responses to simulate real world behaviors. Some benefits to unit testing include:
+
+- Quickly find and isolate bugs in code.
+- Provides documentation for each module of code by way of indicating what certain code blocks should be doing.
+- A helpful gauge that a refactor has gone well. The tests should still pass after code has been refactored.
+
+In the context of Prisma, this generally means testing a function which makes database calls using the Prisma client.
+
+A single test should focus on the data returned by the Prisma client and how the function handles that data, ensuring that it behaves as expected.
```
Comment (mhwelander) [inline content/300-guides/150-testing/100-unit-testing.mdx]:
```suggestion
A single test should focus on the data returned by the Prisma Client and how the function handles that data, ensuring that it behaves as expected.
```

Comment (mhwelander) [reply content/300-guides/150-testing/100-unit-testing.mdx]: Maybe a bit more emphasis on "A single test should focus on how **your function logic** handles different inputs (such as a `null` value or an empty list)."

File: content/300-guides/150-testing/100-unit-testing.mdx [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+
+To ensure your unit tests are isolated from external factors you can mock the Prisma client, this means you get the benefits of being able to use your schema (**_type-safety_**), without having to make actual calls to your database when your tests are run.
+
+This guide will cover two approaches to mocking the client, a singleton instance and dependency injection. Both have their merits depending on your use cases. To help with mocking the client the [`jest-mock-extended`](https://github.com/marchaos/jest-mock-extended) package will be used.
+
+```terminal
+npm install jest-mock-extended --save-dev
+```
+
+### Singleton
```
Comment (mhwelander) [inline content/300-guides/150-testing/100-unit-testing.mdx:70]: Nevermind!

File: content/300-guides/150-testing/100-unit-testing.mdx [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+
+1. Create a file at your projects root called `client.ts` and add the following code. This will instantiate a Prisma client instance.
+
+```ts file=client.ts
```
Comment (mhwelander) [inline content/300-guides/150-testing/100-unit-testing.mdx]: Needs indenting - I think this is the case further down as well

File: content/300-guides/150-testing/100-unit-testing.mdx [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+
+```prisma file=schema.prisma
+model User {
+  id    Int     @id @default(autoincrement())
+  email String  @unique
+  name  String?
+}
+```
+
+The following unit tests will mock the process of
```
Comment (mhwelander) [inline content/300-guides/150-testing/100-unit-testing.mdx]:
```suggestion
The following unit tests mock the process of:
```

File: content/300-guides/150-testing/100-unit-testing.mdx (hunk 4 above)
Comment (mhwelander) [inline content/300-guides/150-testing/100-unit-testing.mdx:70]: Introduce - "To mock the client with a singleton" or similar

File: content/300-guides/150-testing/100-unit-testing.mdx [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+  mockReset(prismaMock)
+})
+
+export const prismaMock = (prisma as unknown) as MockProxy<PrismaClient>
+```
+
+The singleton file tells Jest to mock a default export (the Prisma client instantiated in `./client.ts`), and uses the `mockDeep` method from `jest-mock-extended` to enable access to the objects and methods available on the Prisma client. It then resets the mocked instance before each test is run.
+
+### Dependency injection
+
```
Comment (mhwelander) [inline content/300-guides/150-testing/100-unit-testing.mdx]: Introduce - "To mock the client with dependency injection"

Also needs steps so that it matches the singleton examples

File: content/300-guides/150-testing/100-unit-testing.mdx [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+The **_dependency injection_** example passes the context through to the function that is being tested as well as using it to call the mock implementation.
+
+The **_singleton_** example uses the singleton client instance to call the mock implementation.
+
+<TabbedContent tabs={[<FileWithIcon text="with-singleton.ts" icon="file"/>, <FileWithIcon text="with-dependency-injection" icon="file"/>]}>
+
+<tab>
+
+```ts
```
Comment (mhwelander) [inline content/300-guides/150-testing/100-unit-testing.mdx]: I wonder if we can extend these examples to test some _function-level logic_ rather than database logic. Right now it looks like we are testing that `user.update` and `user.create` work, which will always succeed. Why don't we test if the user has checked the `acceptTermsAndConditions` checkbox (should create user) vs if not (should show error message)? that's application-level logic rather than database-level.

File: content/300-guides/150-testing/150-integration-testing.mdx [hunk 9]
```diff
@@ -0,0 +1,450 @@
...
+
+<TopBlock>
+
+Integration tests focus on testing how separate parts of the program work together. In the context of applications using a database, integration tests usually require a database to be available and contain data that is convenient to the scenarios intended to be tested.
+
+One way to simulate a real world environment is to use [Docker](https://www.docker.com/get-started) to encapsulate a database and some test data. This can be spun up and torn down with the tests and so operate as an isolated environment away from your production databases.
+
+</TopBlock>
+
+#### Prerequisites
```
Comment (mhwelander) [inline content/300-guides/150-testing/150-integration-testing.mdx]:
```suggestion
## Prerequisites
```

Semantically incorrect to write ####

File: content/300-guides/150-testing/150-integration-testing.mdx [hunk 10]
```diff
@@ -0,0 +1,450 @@
...
+}
+```
+
+</details>
+
+The guide uses a singleton pattern for Prisma Client setup. Refer to the [singleton](./unit-testing#singleton) docs for a walk through of how to set that up.
+
+## Add Docker to your project
+
+With Docker and Docker compose both installed on your machine, begin by creating a `docker-compose.yml` file at your projects root. Here you will add a Postgres image and specify the environments credentials.
```
Comment (mhwelander) [inline content/300-guides/150-testing/150-integration-testing.mdx]: General comment: Would make things into steps if they are steps, hard to see what part of each section is a task and what is extra info

File: content/300-guides/150-testing/150-integration-testing.mdx [hunk 11]
```diff
@@ -0,0 +1,450 @@
...
+      - '5433:5432'
+    environment:
+      POSTGRES_USER: prisma
+      POSTGRES_PASSWORD: prisma
+      POSTGRES_DB: tests
+```
+
+> **Note**: The version used here is the latest at the time of writing, if you are following along be sure to use the same version for consistency.
+
+Breaking down the file:
```
Comment (mhwelander) [inline content/300-guides/150-testing/150-integration-testing.mdx]: Something more formal - e.g. "The `docker-compoose.yml` file defines the following:"

File: content/300-guides/150-testing/150-integration-testing.mdx [hunk 12]
```diff
@@ -0,0 +1,450 @@
...
+
+- The Postgres image is specified with the (version) 13 tag. This will be downloaded if you do not have it locally available.
```
Comment (mhwelander) [inline content/300-guides/150-testing/150-integration-testing.mdx]:
```suggestion
- The Postgres image (`postgres`) and version tag (`:13`). This will be downloaded if you do not have it locally available.
```

File: content/300-guides/150-testing/150-integration-testing.mdx [hunk 13]
```diff
@@ -0,0 +1,450 @@
...
+- The port `5433` is mapped to the internal (Postgres default) port `5432`. This will be the port number the database is exposed on externally.
+- The database user credentials are set and the database given a name.
+
+To connect to the database in the container you will need to create a new connection string. Given the credentials specified in the compose file that would look like the following:
```
Comment (mhwelander) [inline content/300-guides/150-testing/150-integration-testing.mdx]:
```suggestion
To connect to the database in the container, create a new connection string with the credentials defined in the `docker-compose.yml` file. For example:
```

Simplified language

File: content/300-guides/150-testing/150-integration-testing.mdx [hunk 14]
```diff
@@ -0,0 +1,450 @@
...
+(4 rows)
+```
+
+</CmdResult>
+
+</CodeWithResult>
+
+## Add a seed script
+
+Your integration tests will be run against a database in a test environment instead of your actual production or development environments.
```
Comment (mhwelander) [inline content/300-guides/150-testing/150-integration-testing.mdx]:
```suggestion
Integration tests run against a database in a dedicated test environment instead of the production or development environments.
```

File: content/300-guides/150-testing/150-integration-testing.mdx [hunk 15]
```diff
@@ -0,0 +1,450 @@
...
+
+For the tests to best simulate real world scenarios you can seed the database on startup, thus providing your tests with some data to test against.
```
Comment (mhwelander) [inline content/300-guides/150-testing/150-integration-testing.mdx]: I would emphasize realistic data more and cut 'thus' - e.g. "You can use a seed script to pre-populate the database with real world test data, such as names with special characters, and easily run your tests against the same dataset multiple times."

File: content/300-guides/150-testing/150-integration-testing.mdx [hunk 16]
```diff
@@ -0,0 +1,450 @@
...
+    "migrate": "npx prisma migrate dev --name init",
+    "docker:up": "docker-compose up -d",
+    "docker:down": "docker compose down",
+    "seed": "node --loader ts-node/esm ./prisma/seed.ts",
+    "bootstrap": "yarn docker:up && yarn migrate && yarn seed",
+    "test": "yarn bootstrap && jest -i && yarn docker:down"
+  },
+```
+
+To run create the container and run the test you would run the `test` script. Breaking this down:
```
Comment (mhwelander) [inline content/300-guides/150-testing/150-integration-testing.mdx]: Confused about this sentence: To run create the container and run the test you would run the `test` script. Should it be "To start the test, run the `test` script" and if so can we add:

```
npm run test
```

Would replace 'breaking this down' to something more formal - "The `seed` script runs the following steps:" or just "The `seed` script" ... "creates a container", "applies a migration"

File: content/300-guides/150-testing/150-integration-testing.mdx [hunk 17]
```diff
@@ -0,0 +1,450 @@
...
+
+1. `docker compose up -d` is run to create the container with the Postgres image, creating the database.
```
Comment (mhwelander) [inline content/300-guides/150-testing/150-integration-testing.mdx]: This is passive - if we change the intro sentence to "The `seed` script:" we can make the steps more active:

1) runs `docker compuse up -d` to create the container (...)
2) Runs the `seed.ts` script to create some initial data

**Comment by mhwelander** [comment]
💯 Added some comments and suggestions, I really like the general explanation followed by a concrete example!

File: content/300-guides/150-testing/100-unit-testing.mdx (hunk 5 above)
Comment (molebox) [reply content/300-guides/150-testing/100-unit-testing.mdx]: As soon as i hit save it auto formats it back. I need to figure this out cos its causing lots of comments

File: content/300-guides/150-testing/150-integration-testing.mdx (hunk 15 above)
Comment (molebox) [reply content/300-guides/150-testing/150-integration-testing.mdx]: Argh, i cant believe i used 'thus'. Shoot me next time please

File: content/300-guides/150-testing/150-integration-testing.mdx [hunk 18]
```diff
@@ -1,13 +1,458 @@
...
+      POSTGRES_PASSWORD: prisma
+      POSTGRES_DB: tests
+```
+
+> **Note**: The compose version used here (`3.9`) is the latest at the time of writing, if you are following along be sure to use the same version for consistency.
+
+The `docker-compose.yml` file defines the following:
+
+- The Postgres image (`postgres`) and version tag (`:13`). This will be downloaded if you do not have it locally available.
+- The port `5433` is mapped to the internal (Postgres default) port `5432`. This will be the port number the database is exposed on externally.
```
Comment (Jolg42) [inline content/300-guides/150-testing/150-integration-testing.mdx:112]: Why not mapping both to `5432`?

Comment (molebox) [reply content/300-guides/150-testing/150-integration-testing.mdx:112]: I had issues with this and found that chaging the port solved it. I was going to document that but i cant find the reason why it didnt work. I though people might ask that exact question though....

File: content/300-guides/150-testing/150-integration-testing.mdx [hunk 19]
```diff
@@ -1,13 +1,458 @@
...
+This setup isolates a real world scenario so that you can test your applications functionality against real data in a controlled environment. 
+
+You can add some scripts to your projects `package.json` file which will setup the database and run the tests, then afterwards manually destroy the container.
+
+```json file=package.json
+  "scripts": {
+    "migrate": "npx prisma migrate dev --name init",
+    "docker:up": "docker-compose up -d",
+    "docker:down": "docker compose down",
+    "test": "yarn docker:up && yarn migrate && jest -i"
```
Comment (Jolg42) [inline content/300-guides/150-testing/150-integration-testing.mdx]:
```suggestion
    "migrate:init": "npx prisma migrate dev --name init",
    "docker:up": "docker-compose up -d",
    "docker:down": "docker compose down",
    "test": "yarn docker:up && yarn migrate:init && jest --runInBand"
```

Maybe?

File: content/300-guides/150-testing/150-integration-testing.mdx [hunk 20]
```diff
@@ -1,13 +1,458 @@
...
+  },
+```
+
+The `test` script does the following:
+
+1. Runs `docker compose up -d` to create the container with the Postgres image and database.
+1. Runs a migration on the database using your projects schema, this creates the tables in the containers database.
```
Comment (Jolg42) [inline content/300-guides/150-testing/150-integration-testing.mdx]:
```suggestion
1. Runs a migration on the database using your project's schema, this creates the tables in the container's database.
```

File: content/300-guides/150-testing/150-integration-testing.mdx [hunk 21]
```diff
@@ -1,13 +1,458 @@
...
+      },
+    },
+  });
+
+  // Expect the new customer to have been created and match the input
+  expect(newCustomer).toEqual(customer);
+  // Expect the new order to have been created and contain the new customer
+  expect(newOrder).toHaveProperty("customerId", 2);
+
+  done()
```
Comment (Jolg42) [inline content/300-guides/150-testing/150-integration-testing.mdx]: Just curious, is it required to use `done()`? 
I think it should work without (and removing it from function arg)?

File: content/300-guides/150-testing/150-integration-testing.mdx (hunk 19 above)
Comment (molebox) [reply content/300-guides/150-testing/150-integration-testing.mdx]: I can add the `:init`. `-i` is short for `--runInBand` See: https://jestjs.io/docs/next/cli#--runinband

File: content/300-guides/150-testing/150-integration-testing.mdx (hunk 21 above)
Comment (molebox) [reply content/300-guides/150-testing/150-integration-testing.mdx]: Its not required, in fact i think i can removbe that now. I added it as i had a hanging async operation

File: content/300-guides/150-testing/150-integration-testing.mdx (hunk 18 above)
Comment (Jolg42) [reply content/300-guides/150-testing/150-integration-testing.mdx:112]: I think it only helps if you already have a PostgreSQL database service started locally, Then the 5432 port is already occupied and Docker cannot use it.

File: content/300-guides/150-testing/150-integration-testing.mdx (hunk 19 above)
Comment (Jolg42) [reply content/300-guides/150-testing/150-integration-testing.mdx]: Yes! -i is nice and short but hard to remember what it does without reading the doc 😅 where the longer one is nice to understand I think, anyway it's in a script so it's the same length to type the command 😄

File: content/300-guides/150-testing/150-integration-testing.mdx [hunk 22]
```diff
@@ -1,13 +1,458 @@
...
+    email: "harry@hogwarts.io",
+  };
+  // The new orders details
+  const order: OrderInput = {
+    customer,
+    productId: 3,
+    quantity: 1,
+  };
+
+  // The productId supplied doesnt exit so the function should return an "Out of stock" message
```
Comment (Jolg42) [inline content/300-guides/150-testing/150-integration-testing.mdx]:
```suggestion
  // The productId supplied doesn't exit so the function should return an "Out of stock" message
```

File: content/300-guides/150-testing/150-integration-testing.mdx (hunk 20 above)
Comment (molebox) [reply content/300-guides/150-testing/150-integration-testing.mdx]: Im unsure about this. For me the project is a possesive noun (it belongs to the user) so it doesn't need an apostrophe. The container should be changed though.

Ive forgotten alot of the rules, was a while ago i studied english! Will have a google around for more info. I could be wrong

Comment (molebox) [reply content/300-guides/150-testing/150-integration-testing.mdx]: I went with "Runs a migration on the database using your projects schema, this creates the tables in the container's database."

File: content/300-guides/150-testing/150-integration-testing.mdx (hunk 18 above)
Comment (molebox) [reply content/300-guides/150-testing/150-integration-testing.mdx:112]: Ohhh. Hmm. But i dont have anything else running on that port, i checked. Lemme double check again now and change port. It would make much more sense to map to the same port as thats what people expect too

Comment (molebox) [reply content/300-guides/150-testing/150-integration-testing.mdx:112]: Yep, it doesnt work for me. I cant authenticate

Comment (Jolg42) [reply content/300-guides/150-testing/150-integration-testing.mdx:112]: Docker compose logs probably show something about the port if it's already used / cannot be used

Comment (molebox) [reply content/300-guides/150-testing/150-integration-testing.mdx:112]: Nope, not that i can see at least

```
$ docker-compose logs -f -t
Attaching to integration-tests-prisma
integration-tests-prisma | 2021-05-12T09:27:28.866263300Z The files belonging to this database system will be owned by user "postgres".
integration-tests-prisma | 2021-05-12T09:27:28.866310600Z This user must also own the server process.
integration-tests-prisma | 2021-05-12T09:27:28.866314000Z
integration-tests-prisma | 2021-05-12T09:27:28.866316600Z The database cluster will be initialized with locale "en_US.utf8".
integration-tests-prisma | 2021-05-12T09:27:28.866319000Z The default database encoding has accordingly been set to "UTF8".
integration-tests-prisma | 2021-05-12T09:27:28.866321400Z The default text search configuration will be set to "english".
integration-tests-prisma | 2021-05-12T09:27:28.866323500Z
integration-tests-prisma | 2021-05-12T09:27:28.866325500Z Data page checksums are disabled.
integration-tests-prisma | 2021-05-12T09:27:28.866327500Z
integration-tests-prisma | 2021-05-12T09:27:28.866329400Z fixing permissions on existing directory /var/lib/postgresql/data ... ok
integration-tests-prisma | 2021-05-12T09:27:28.867173800Z creating subdirectories ... ok
integration-tests-prisma | 2021-05-12T09:27:28.867261600Z selecting dynamic shared memory implementation ... posix
integration-tests-prisma | 2021-05-12T09:27:28.878419700Z selecting default max_connections ... 100
integration-tests-prisma | 2021-05-12T09:27:28.892358200Z selecting default shared_buffers ... 128MB
integration-tests-prisma | 2021-05-12T09:27:28.906787900Z selecting default time zone ... Etc/UTC
integration-tests-prisma | 2021-05-12T09:27:28.908146800Z creating configuration files ... ok
integration-tests-prisma | 2021-05-12T09:27:29.018818300Z running bootstrap script ... ok
integration-tests-prisma | 2021-05-12T09:27:29.350018100Z performing post-bootstrap initialization ... ok
integration-tests-prisma | 2021-05-12T09:27:29.685673500Z syncing data to disk ... ok
integration-tests-prisma | 2021-05-12T09:27:29.685710500Z
integration-tests-prisma | 2021-05-12T09:27:29.685714600Z
integration-tests-prisma | 2021-05-12T09:27:29.685717100Z Success. You can now start the database server using:
integration-tests-prisma | 2021-05-12T09:27:29.685719400Z
integration-tests-prisma | 2021-05-12T09:27:29.685721500Z     pg_ctl -D /var/lib/postgresql/data -l logfile start
integration-tests-prisma | 2021-05-12T09:27:29.685723700Z
integration-tests-prisma | 2021-05-12T09:27:29.685676500Z initdb: warning: enabling "trust" authentication for local connections
integration-tests-prisma | 2021-05-12T09:27:29.685731400Z You can change this by editing pg_hba.conf or using the option -A, or
integration-tests-prisma | 2021-05-12T09:27:29.685733500Z --auth-local and --auth-host, the next time you run initdb.
integration-tests-prisma | 2021-05-12T09:27:29.720693100Z waiting for server to start....2021-05-12 09:27:29.720 UTC [47] LOG:  starting PostgreSQL 13.2 (Debian 13.2-1.pgdg100+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 8.3.0-6) 8.3.0, 64-bit
integration-tests-prisma | 2021-05-12T09:27:29.727427800Z 2021-05-12 09:27:29.727 UTC [47] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
integration-tests-prisma | 2021-05-12T09:27:29.750927700Z 2021-05-12 09:27:29.750 UTC [48] LOG:  database system was shut down at 2021-05-12 09:27:29 UTC
integration-tests-prisma | 2021-05-12T09:27:29.759384400Z 2021-05-12 09:27:29.759 UTC [47] LOG:  database system is ready to accept connections
integration-tests-prisma | 2021-05-12T09:27:29.794525900Z  done
integration-tests-prisma | 2021-05-12T09:27:29.794544600Z server started
integration-tests-prisma | 2021-05-12T09:27:30.064517200Z CREATE DATABASE
integration-tests-prisma | 2021-05-12T09:27:30.065018600Z
integration-tests-prisma | 2021-05-12T09:27:30.065259400Z
integration-tests-prisma | 2021-05-12T09:27:30.065299300Z /usr/local/bin/docker-entrypoint.sh: ignoring /docker-entrypoint-initdb.d/*
integration-tests-prisma | 2021-05-12T09:27:30.065302600Z
integration-tests-prisma | 2021-05-12T09:27:30.066200200Z 2021-05-12 09:27:30.066 UTC [47] LOG:  received fast shutdown request
integration-tests-prisma | 2021-05-12T09:27:30.073528400Z waiting for server to shut down....2021-05-12 09:27:30.073 UTC [47] LOG:  aborting any active transactions
integration-tests-prisma | 2021-05-12T09:27:30.074529600Z 2021-05-12 09:27:30.074 UTC [47] LOG:  background worker "logical replication launcher" (PID 54) exited with exit code 1
integration-tests-prisma | 2021-05-12T09:27:30.074697000Z 2021-05-12 09:27:30.074 UTC [49] LOG:  shutting down
integration-tests-prisma | 2021-05-12T09:27:30.116218600Z 2021-05-12 09:27:30.116 UTC [47] LOG:  database system is shut down
integration-tests-prisma | 2021-05-12T09:27:30.166455500Z  done
integration-tests-prisma | 2021-05-12T09:27:30.166491000Z server stopped
integration-tests-prisma | 2021-05-12T09:27:30.166639300Z
integration-tests-prisma | 2021-05-12T09:27:30.166648100Z PostgreSQL init process complete; ready for start up.
integration-tests-prisma | 2021-05-12T09:27:30.166650900Z
integration-tests-prisma | 2021-05-12T09:27:30.194092700Z 2021-05-12 09:27:30.193 UTC [1] LOG:  starting PostgreSQL 13.2 (Debian 13.2-1.pgdg100+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 8.3.0-6) 8.3.0, 64-bit
integration-tests-prisma | 2021-05-12T09:27:30.194232400Z 2021-05-12 09:27:30.194 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
integration-tests-prisma | 2021-05-12T09:27:30.194335400Z 2021-05-12 09:27:30.194 UTC [1] LOG:  listening on IPv6 address "::", port 5432
integration-tests-prisma | 2021-05-12T09:27:30.208729200Z 2021-05-12 09:27:30.208 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
integration-tests-prisma | 2021-05-12T09:27:30.224341600Z 2021-05-12 09:27:30.224 UTC [75] LOG:  database system was shut down at 2021-05-12 09:27:30 UTC
integration-tests-prisma | 2021-05-12T09:27:30.234287700Z 2021-05-12 09:27:30.234 UTC [1] LOG:  database system is ready to accept connections
```

Comment (molebox) [reply content/300-guides/150-testing/150-integration-testing.mdx:110]: Thats me starting it on 5433 as i cant start it on 5432
